### ISSUE TITLE
Migrate Dockerfile used to build Dapr to static, nonroot runtime image

### ISSUE THREAD

**Issue by mchmarny** [dapr/dapr#1551]
The current Dockerfile used to build Dapr uses `alpine:latest` as base image during build and `scratch` as the runtime. Proposal:

### Use go versioned based image

```yaml
FROM golang:1.14.2 as builder
```

vs

```yaml
FROM alpine:latest as alpine
```
### Use `static:nonroot` and `distroless` image in runtime**

"Distroless" images do not contain package managers, shells or any other programs. 

```yaml
FROM gcr.io/distroless/static:nonroot
```

vs 

```yaml
FROM scratch
```

> If pulling from `gcr.io` is an issue we can find alt

### Benefits 

* Minimize layers to maximize build and pull performance (images small)
* Minimize entire runtime environment

### References 

* https://github.com/GoogleContainerTools/distroless)
* https://medium.com/@dwdraju/distroless-is-for-security-if-not-for-size-6eac789f695f

### Updated Dockerfile 

```yaml
FROM FROM golang:1.14.2 as builder

ARG TARGETPLATFORM
ADD . /dist

RUN if [ $TARGETPLATFORM = linux/amd64 ]; then mkdir -p /dist/linux/amd64 && mv /dist/linux_amd64/release /dist/linux/amd64/; fi
RUN if [ $TARGETPLATFORM = linux/arm/v7 ]; then mkdir -p /dist/linux/arm/v7 && mv /dist/linux_arm/release /dist/linux/arm/v7/; fi

FROM gcr.io/distroless/static:nonroot

ARG TARGETPLATFORM
ARG PKG_FILES
ENTRYPOINT []
WORKDIR /

COPY --from=builder /dist/$TARGETPLATFORM/release/$PKG_FILES /
```

**Comment by youngbupark** [comment]
We have used [scratch base image](https://hub.docker.com/_/scratch/)[~0B] for the final image.

As we know, dapr dockerfile uses multi-stage build. alpine image is used to copy ca and the target arch binary from local disk to container. all dapr binaries are statically linked. 

scratch base image(0MB) is smaller than distroless(1.83MB), but it makes sense to switch to distroless in terms of security manner, not size of image.

cc/ @yaron2 @amanbha @msfussell @orizohar

**Comment by youngbupark** [comment]
@yaron2 @amanbha @yolember no objections to switch to `gcr.io/distroless/static:nonroot` ?

**Comment by mchmarny** [comment]
We should do this early and let it soak to validate the vendored libs not doing something unnatural

**Comment by tcnghia** [comment]
`distroless` just added `arm64`. See gcr.io/distroless/static:nonroot (thanks @mattmoor for https://github.com/GoogleContainerTools/distroless/pull/586/files )

**Comment by mattmoor** [comment]
Wasn't aware of that blog post, but you can watch my original talk on it at the top of the distroless README.  Hopefully entertaining, and only slightly dated (I am no longer drinking the Bazel kool-aid 😉 ).

Happy to swing by and chat if it'd be helpful?

**Comment by artursouza** [comment]
This is the current content for `Dockerfile`:

```
FROM alpine:latest as alpine
RUN apk add -U --no-cache ca-certificates
# current directory must be ./dist

FROM gcr.io/distroless/static:nonroot
ARG PKG_FILES
WORKDIR /
COPY --from=alpine /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/
COPY /$PKG_FILES /
```

It does not make sense to fix the golang version because building happens outside of the container and golang is not used at all. The current version already uses nonroot distroless base image for runtime. Closing this issue.

**Comment by artursouza** [comment]
Fixed as part of: #2536
